### DISCUSSION TITLE
one to one self relation

### DISCUSSION THREAD

**Discussion by daspete** [prisma/prisma#11743] [Q&A] [answered]
Hello there,

i just stumbled upon a strange syntax when i try to make a typical User model for a management system, where users can create other users and we want to track who created a user. so i would write a model like this

```
model User {
    id              String      @id @default(dbgenerated()) @map("_id") @db.ObjectId
    createdAt       DateTime
    updatedAt       DateTime
    createdById     String?     @db.ObjectId
    createdBy       User?       @relation("UserCreatedBy", fields: [createdById], references: [id], onDelete: NoAction, onUpdate: NoAction)
}
```

but this is throwing a validation error: 
```
Error validating field `createdBy` in model `User`: The relation field `createdBy` on Model `User` is missing an opposite relation field on the model `User`. Either run `prisma format` or add it manually.
```

and when i write:

```
model User {
    id              String      @id @default(dbgenerated()) @map("_id") @db.ObjectId
    createdAt       DateTime
    updatedAt       DateTime
    createdById     String?     @db.ObjectId
    createdBy       User?       @relation("UserCreatedBy", fields: [createdById], references: [id], onDelete: NoAction, onUpdate: NoAction)
    createdBy2      User?       @relation("UserCreatedBy")
}
```

It's working...

My question, why is it necessary to make the "createdBy2" field?

**Comment by Luke265** [comment, marked as answer]
I think it is just because of the design, opposite relation fields are required. Anyway, `where users can create other users` so this should be one to many relation

```
model User {
    id              String      @id @default(dbgenerated()) @map("_id") @db.ObjectId
    createdAt       DateTime
    updatedAt       DateTime
    createdById     String?     @db.ObjectId
    createdBy       User?       @relation("UserCreatedBy", fields: [createdById], references: [id], onDelete: NoAction, onUpdate: NoAction)
    createdUsers    User[]      @relation("UserCreatedBy")
}
```
then you will be able to find all users created by specific user using either of these two approaches:
```js
const user = prisma.user.findMany({
    include: {
        createdUsers: true
    }
    where: {
        id: userId
    }
});
console.log(user.createdUsers);
// or
const users = prisma.user.findMany({
    where: {
        createdById: {
            eq: userId
        }
    }
});
console.log(users);
```

or if you still want one to one relation, then just change `createdBy2` to less confusing name like `createdUser`.

**Reply by daspete** [reply]
thank you for your answer, i understand that, for this one use case, it sounds like a solution... so, you always need to define at least two relations in a model to get things up and running, right? but just for the self relations? i just wonder, if there isn't a way to automatically add a hidden self relation, when there isn't defined a second one. But for now... i can life with that.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
